### ISSUE TITLE
false negative when CPE vendor is star

### ISSUE THREAD

**Issue by CedricSaf** [DependencyTrack/dependency-track#2580]
### Current Behavior

Using this sbom that contain vendor=**linux**:
```
{
  "bomFormat": "CycloneDX",
  "specVersion": "1.4",
  "version": 1,
  "components": [
    {
      "type": "library",
      "name": "kernel-4.19.139",
      "version": "4.19.139",
      "cpe": "cpe:2.3:o:linux:linux_kernel:4.19.139:*:*:*:*:*:*:*"
    }
  ]
}
```
This CVE is identified: CVE-2020-11494 

Using the same sbom with CPE **vendor=***  :
```
{
  "bomFormat": "CycloneDX",
  "specVersion": "1.4",
  "version": 1,
  "components": [
    {
      "type": "library",
      "name": "kernel-4.19.139",
      "version": "4.19.139",
      "cpe": "cpe:2.3:o:*:linux_kernel:4.19.139:*:*:*:*:*:*:*"
    }
  ]
}
```
The CVE CVE-2020-11494  is NOT identified as a vuln.

In parallel, CVE-2022-47929 is correctly found in both case.

### Steps to Reproduce

Dependency-Track v4.7.0
Build ID: e2ebe7f4-72a1-446b-98ee-5f2b758de8ee
Built On: 16 Dec 2022 at 18:52:12 

Create a project with sbom dumped above and check vulns found.

### Expected Behavior

From my understanding, the CVE-2020-11494 should be found with both CPE (vendor = linux and vendor = *) set in component identity.

### Dependency-Track Version

4.7.0

### Dependency-Track Distribution

Container Image

### Database Server

PostgreSQL

### Browser

Mozilla Firefox

### Checklist

- [X] I have read and understand the [contributing guidelines](https://github.com/DependencyTrack/dependency-track/blob/master/CONTRIBUTING.md#filing-issues)
- [X] I have checked the [existing issues](https://github.com/DependencyTrack/dependency-track/issues) for whether this defect was already reported

**Comment by syalioune** [comment]
Hello @CedricSaf,

If you follow [NIST CPE Matching specification](https://nvlpubs.nist.gov/nistpubs/Legacy/IR/nistir7696.pdf) and especially *Table 6-2*

![image](https://user-images.githubusercontent.com/6144741/226184418-be058aca-e83d-444a-a81d-7a55eca4180e.png)

You fall into case 13 for `vendor` attribute match :

- `Source AV` is the CPE from NVD : `linux`
- `Target AV` is your CPE : `*`
- `linux` is a subset of `*`

That relation is `falsy` because your CPE is too generic. The other way around is OK.

**Comment by syalioune** [comment]
Hello Cedric,
I forgot to mention it but both CVE are not identified on my side which is consistent. `CVE-2022-47929` is however indeed identified if I enable the Internal Analysis Fuzzy Matching (it uses an internal [Lucene index](https://lucene.apache.org/) to deal with NVD CPE data weirdness). Do you have it enabled on your side ?
Don't know yet why Fuzzy Matching returns `CVE-2022-47929` and not `CVE-2020-11494`.

**Comment by syalioune** [comment]
> Don't know yet why Fuzzy Matching returns CVE-2022-47929 and not CVE-2020-11494.

The lucene index query returns [only the first 1000 matching CPEs](https://github.com/DependencyTrack/dependency-track/blob/master/src/main/java/org/dependencytrack/search/FuzzyVulnerableSoftwareSearchManager.java#L168). Chances are that `CVE-2020-11494` is not on the lucky side of the result cutoff.

**Comment by syalioune** [comment]
It depend on what you consider as `source attribute value` and `target attribute value` on `table 6-2` when doing the comparison.
Fuzzy analysis aside, DT always has the following approach :

- `Source` is the CPE configuration from the `NVD`
- `Target` is the CPE from your component

Taking [CVE-2020-11494](https://nvd.nist.gov/vuln/detail/CVE-2020-11494) as an example :

- `Source` is `cpe:2.3:o:linux:linux_kernel:*:*:*:*:*:*:*:*` (Known affected software)
- `Target` is `cpe:2.3:o:linux:linux_kernel:4.19.139:*:*:*:*:*:*:*` (Your CPE)
- Your CPE is a `subset` of the `known affected software`
- The CVE `CVE-2020-11494` is a match

> Most of 670 vuln (exemple: https://github.com/advisories/GHSA-64rp-89cr-p8q9) are a subset of the CPE with vendor=*

For DT : the CPE is a `superset` of the `known affected software` hence no match.

**Comment by Jasper-Ben** [comment]
:wave: @CedricSaf @syalioune we just stumbled into this issue as well.

We are trying to use DT to keep track of components within an embedded Linux system created with Yocto. Unfortunately, the CPE strings provided by Yocto itself are [quite generic](https://github.com/openembedded/openembedded-core/blob/e7c1def3c3c3a72249802ef6fb64292277a7a53e/meta/lib/oe/cve_check.py#L159) (e.g. `cpe:2.3:*:*:linux_kernel:5.15.87:*:*:*:*:*:*:*`) with no easy fix, as this would require structural change to the entire Yocto ecosystem.

This leaves us in a situation where disabling fuzzy CPE matching will result in 0 hits and with fuzzy matching enabled we probably might as well roll a dice to figure out whether we are effected by a CVE or not, as it will result in many false positives and negatives, as it will also fuzz the already sparse "well-known" information we have.

I am therefore wondering, if it makes sense to add another configurable search mechanism in addition to NIST CPE matching and fuzzy matching (e.g. REGEX matching comes to mind)? Or maybe it would help to have the `source attribute value` and `target attribute value` in DT configurable? I still need to fully grasp the NIST CPE matching specification but from your comment I understood that switching source and target might help, if the components CPE information is sparse.

**Comment by Jasper-Ben** [comment]
Generally, I am confused by the results generated by DT.
When disabling fuzzy matching and adding `cpe:2.3:o:linux:linux_kernel:5.15.71:*:*:*:*:*:*:*`, DT will report 388 CVEs.
When searching for the same CPE on NIST directly it will report[ 819 CVEs](https://nvd.nist.gov/vuln/search/results?form_type=Advanced&results_type=overview&isCpeNameSearch=true&seach_type=all&query=cpe:2.3:o:linux:linux_kernel:5.15.71:*:*:*:*:*:*:*). That's over twice(!) the CVEs reported.

**Comment by Jasper-Ben** [comment]
I finally found the time for a thorough read of the NIST specification and things make a lot more sense now (who would have thought, that actually trying to understand something would help... duuuh... :wink:). So yes, DT does work as expected and according to spec.

> Fuzzy analysis aside, DT always has the following approach :
>
>    Source is the CPE configuration from the NVD
>   Target is the CPE from your component

And this is actually the only sensible approach, as the CVEs CPE might contain wildcards (not to be confused with ANY), which could easily lead to undefined relations (table 6-2, cases 4, 8, 11), while wildcards for component CPEs do not make much sense.
So yes, a component CPE MUST (considering minimum CPE matching) be well defined (i.e. no "ANY" for "vital" attributes).

However, the specification leaves for some wiggle-room, which can help with this use-case:

![image](https://github.com/DependencyTrack/dependency-track/assets/12223583/8a113859-8857-4e9f-865e-80003a54676c)

For example, DT could add a configurable option to treat any component CPE attributes that contains an ANY value as a NA (not applicable) value for both source and target CPE. Effectively, this would result in:

CVE CPE:             "cpe:2.3:o:NA:linux_kernel:4.19.139:NA:NA:NA:NA:NA:NA:NA"
Component CPE:  "cpe:2.3:o:NA:linux_kernel:4.19.139:NA:NA:NA:NA:NA:NA:NA"

which would be a match.

I believe this should be easily implemented (if COMPONENT_CPE_ATTRIBUTE == ANY -> COMPONENT_CPE_ATTRIBUTE & CVE_CPE = NA) and would in my opinion even make for a good default setting, as there isn't really a good reason for comparing a component CPE ANY value with a CVE CPE value, due to false negatives.

**Comment by Jasper-Ben** [comment]
For visibility: we are fixing this as part of https://github.com/DependencyTrack/dependency-track/issues/2988.

Thus, feel free to close this issue as duplicate.

**Comment by CedricSaf** [comment]
Duplicate of  https://github.com/DependencyTrack/dependency-track/issues/2988
